### PR TITLE
Timeseries Notes and Slides

### PR DESCRIPTION
**Pull request by trgardos**
Mostly ready for review. There is one more TODO and then review comments to address.

### CODE REVIEW

File: ds701_book/26-TimeSeries.qmd [hunk 1]
```diff
@@ -0,0 +1,953 @@
...
+## Techniques to achieve stationarity
+
+1. **Differencing**: Subtracting the previous observation from the current observation to remove trends and seasonality.
+2. **Transformation**: Applying mathematical transformations such as logarithms or square roots to stabilize the variance.
+3. **Decomposition**: Separating the time series into trend, seasonal, and residual components to analyze and adjust each part individually.
+4. **Smoothing**: Using techniques like moving averages to smooth out short-term fluctuations and highlight longer-term trends.
+
+## Demonstration of stationarity tests
+
+- **ADF Test (Augmented Dickey-Fuller Test)**: A statistical test used to determine if a time series is stationary. It tests the null hypothesis that a unit root is present in the time series.
```
Comment (sladenheim) [inline ds701_book/26-TimeSeries.qmd:739]: What is a unit root?

File: ds701_book/26-TimeSeries.qmd [hunk 2]
```diff
@@ -0,0 +1,953 @@
...
+- $y_t$ is the current value,
+- $c$ is a constant,
+- $\phi_1, \phi_2, \ldots, \phi_p$ are the coefficients,
+- $\epsilon_t$ is the white noise error term.
+
+**Key Points**:
+
+1. **Lag Order (p)**: The number of lagged observations included in the model.
+2. **Stationarity**: The series should be stationary for the AR model to be effective.
+3. **Parameter Estimation**: Methods like Yule-Walker equations or Maximum Likelihood Estimation (MLE) are used to estimate the parameters.
```
Comment (sladenheim) [inline ds701_book/26-TimeSeries.qmd:775]: Are the parameters the coefficients? If yes, we should make that link.

File: ds701_book/26-TimeSeries.qmd [hunk 3]
```diff
@@ -0,0 +1,953 @@
...
+- $y_t$ is the current value,
+- $\mu$ is the mean of the series,
+- $\epsilon_t$ is the white noise error term,
+- $\theta_1, \theta_2, \ldots, \theta_q$ are the coefficients.
+
+**Key Points**:
+
+  1. **Lag Order (q)**: The number of lagged forecast errors included in the model.
+  2. **Stationarity**: The series should be stationary for the MA model to be effective.
+  3. **Parameter Estimation**: Methods like Maximum Likelihood Estimation (MLE) are used to estimate the parameters.
```
Comment (sladenheim) [inline ds701_book/26-TimeSeries.qmd:797]: Same here for linking coefficients to parameters.

File: ds701_book/26-TimeSeries.qmd [hunk 4]
```diff
@@ -0,0 +1,953 @@
...
+
+## Demonstration of stationarity tests using Python
+
+```{.python}
+from statsmodels.tsa.stattools import adfuller, kpss, pp_test
+```
+
+TODO: Add code to demonstrate the stationarity tests.
+
+# Time Series Models
```
Comment (sladenheim) [inline ds701_book/26-TimeSeries.qmd:751]: On this slide, I think it would be helpful to state what we are trying to do with Time Series Models. Using time series data we construct time series models in order to make future predictions.

File: ds701_book/26-TimeSeries.qmd [hunk 5]
```diff
@@ -0,0 +1,953 @@
...
+  - **Blocked Cross-Validation**: The data is divided into blocks, and each block is used as a test set while the preceding blocks are used for training.
+
+## Key Points:
+1. **Respect Temporal Order**: Ensure that the training set always precedes the test set to avoid data leakage.
+2. **Multiple Techniques**: Various techniques like rolling forecast origin, time series split, and blocked cross-validation can be used.
+
+## Forecasting and confidence intervals
+
+- **Forecasting**: The process of making predictions about future values based on historical data. Common methods include:
+  - **ARIMA (AutoRegressive Integrated Moving Average)**: A popular time series forecasting method that combines autoregression, differencing, and moving average components.
```
Comment (sladenheim) [inline ds701_book/26-TimeSeries.qmd:904]: Is ARIMA the only model that is used for forecasting? Or do the other methods we introduced also work?

File: ds701_book/26-TimeSeries.qmd [hunk 6]
```diff
@@ -0,0 +1,953 @@
...
+plt.show()
+```
+
+
+
+# Stationarity and Differencing
+
+## Definition of stationarity
+
+A time series is said to be stationary if its statistical properties such as mean, variance, and autocorrelation are constant over time.
```
Comment (sladenheim) [inline ds701_book/26-TimeSeries.qmd:713]: Should we add a definition of autocorrelation somewhere?

File: ds701_book/26-TimeSeries.qmd [hunk 7]
```diff
@@ -0,0 +1,953 @@
...
+plt.figure(figsize=(8, 4))
+irregular_ts.plot(label='Irregular', color='red')
+plt.ylabel('Irregular')
+plt.title('Irregular Component')
+plt.legend()
+plt.show()
+```
+
+**Question**: What do you observe about the irregular component? Is it truly
+random?
```
[the remark this answers is not in the document]
Comment (trgardos) [reply ds701_book/26-TimeSeries.qmd:572]: correct.

File: ds701_book/26-TimeSeries.qmd [hunk 8]
```diff
@@ -0,0 +1,1114 @@
...
+plt.show()
+```
+
+## SARIMAX Model
+
+Now let's create an instance of a **SARIMAX** model -- <b>S</b>easonal 
+<b>A</b>utoregressive <b>I</b>ntegrated <b>M</b>oving <b>A</b>verage with 
+e<b>X</b>ogenous regressors.
+
+It extends the SARIMA model to support exogenous variables.
```
Comment (sladenheim) [inline ds701_book/26-TimeSeries.qmd:932]: What exogenous variables are used in this example?

File: ds701_book/26-TimeSeries.qmd [hunk 9]
```diff
@@ -0,0 +1,1114 @@
...
+- **KPSS Test (Kwiatkowski-Phillips-Schmidt-Shin Test)**: Another test for stationarity that tests the null hypothesis that the time series is stationary around a deterministic trend.
+- **PP Test (Phillips-Perron Test)**: Similar to the ADF test but uses a non-parametric method to account for serial correlation in the error terms.
+
+## Demonstration of stationarity tests using Python
+
+```{.python}
+from statsmodels.tsa.stattools import adfuller, kpss, pp_test
+```
+
+TODO: Add code to demonstrate the stationarity tests.
```
Comment (sladenheim) [inline ds701_book/26-TimeSeries.qmd:768]: One remaining TODO (I understand you may still be working on it).

**Comment by trgardos** [comment]
Mostly updated for slides. Have 1 more TODO to add and then of course the rest of the review comments.

File: ds701_book/26-TimeSeries.qmd [hunk 10]
```diff
@@ -0,0 +1,953 @@
...
+  - $Y(t) = T(t) + S(t) + e(t)$
+  - Where $Y(t)$ is the observed value, $T(t)$ is the trend component, $S(t)$ is the seasonal component, and $e(t)$ is the noise or error term.
+- **Multiplicative Model**: Assumes that the components of the time series are multiplied together. The model can be represented as:
+  - $Y(t) = T(t) \times S(t) \times e(t)$
+  - Where $Y(t)$ is the observed value, $T(t)$ is the trend component, $S(t)$ is the seasonal component, and $e(t)$ is the noise or error term.
+- **Examples**:
+  - Additive: Monthly sales data where the seasonal effect is constant over time.
+  - Multiplicative: Monthly sales data where the seasonal effect increases or decreases proportionally with the trend.
+
+## "Classical" Decomposition
```
[the remark this answers is not in the document]
Comment (sladenheim) [reply ds701_book/26-TimeSeries.qmd:464]: Maybe classical decomposition should come before additive and multiplicative?

File: ds701_book/26-TimeSeries.qmd [hunk 11]
```diff
@@ -0,0 +1,953 @@
...
+
+We'll show a few examples of time series plots and include the code to generate them.
+
+## Time Series Plot
+
+We're going to use a dataset of air passengers per month from 1949 to 1960.
+
+```{python}
+#| echo: false
+path = './air_passengers_1949_1960.csv'
```
[the remark this answers is not in the document]
Comment (trgardos) [reply ds701_book/26-TimeSeries.qmd:240]: done.

File: ds701_book/26-TimeSeries.qmd (hunk 10 above)
[the remark this answers is not in the document]
Comment (trgardos) [reply ds701_book/26-TimeSeries.qmd:464]: Classical and STL support both. Added clarifications.

File: ds701_book/26-TimeSeries.qmd [hunk 12]
```diff
@@ -0,0 +1,953 @@
...
+plt.show()
+
+```
+
+
+
+## Seasonal and Trend decomposition using Loess (STL)
+
+- **STL** is a more flexible and robust method for decomposing time series data.
+- It uses locally weighted regression (Loess) to estimate the trend and seasonal components [@cleveland1990stl].
```
[the remark this answers is not in the document]
Comment (trgardos) [reply ds701_book/26-TimeSeries.qmd:613]: Sure. I added a section, courtesy of gpt-4o, to include in the course notes. I also cited gpt-4o this time since it wrote the whole section.

File: ds701_book/26-TimeSeries.qmd (hunk 6 above)
Comment (trgardos) [reply ds701_book/26-TimeSeries.qmd:732]: Good point. I added a few slides and example to define it.

File: ds701_book/26-TimeSeries.qmd (hunk 1 above)
Comment (trgardos) [reply ds701_book/26-TimeSeries.qmd:758]: I added a definition as footnote. But maybe we should just get rid of this slide? Because we don't really go into any details? or maybe just say these tests exist but we won't go into them.

Comment (trgardos) [reply ds701_book/26-TimeSeries.qmd:758]: Ok. I added examples of applying the tests. We can always skip in lecture and leave in course notes.

File: ds701_book/26-TimeSeries.qmd [hunk 13]
```diff
@@ -0,0 +1,1351 @@
...
+- **Advantages**:
+  - Handles any type of seasonality (e.g., weekly, monthly).
+  - Can deal with missing values and outliers.
+  - Provides smooth and adaptable trend and seasonal components.
+
+> See [Course Notes](#locally-estimated-scatterplot-smoothing----loessgpt4o) for a more detailed description of Loess.
```
Comment (sladenheim) [inline ds701_book/26-TimeSeries.qmd:714]: This links only to a local file on my machine and not to the course webpage.

File: ds701_book/26-TimeSeries.qmd [hunk 14]
```diff
@@ -0,0 +1,1351 @@
...
+plt.plot(data['Number of Passengers'], label='Observed')
+plt.plot(forecast_index, forecast_values, label='Forecast', color='red')
+plt.fill_between(forecast_index, confidence_intervals.iloc[:, 0], confidence_intervals.iloc[:, 1], color='pink', alpha=0.3)
+plt.legend()
+plt.show()
+```
+
+The light pink area shows the 95% confidence interval for the forecast.
+
+**Questions**: what observations do you have about the prediction?
```
Comment (sladenheim) [inline ds701_book/26-TimeSeries.qmd:1258]: This is a single question

File: ds701_book/26-TimeSeries.qmd [hunk 15]
```diff
@@ -0,0 +1,1351 @@
...
+
+
+# Recap and References
+
+
+## Recap
+
+We covered the following key points:
+
+1. **Components of Time Series**:
```
Comment (sladenheim) [inline ds701_book/26-TimeSeries.qmd:1332]: Can remove all colons after this.

File: ds701_book/26-TimeSeries.qmd [hunk 16]
```diff
@@ -0,0 +1,1351 @@
...
+## Key Points:
+1. **Respect Temporal Order**: Ensure that the training set always precedes the test set to avoid data leakage.
+2. **Multiple Techniques**: Various techniques like rolling forecast origin, time series split, and blocked cross-validation can be used.
+
+## Forecasting and confidence intervals
+
+- **Forecasting**: The process of making predictions about future values based on historical data. Common methods include:
+  - **ARIMA (AutoRegressive Integrated Moving Average)**: A popular time series forecasting method that combines autoregression, differencing, and moving average components.
+  - **Exponential Smoothing**: A technique that applies decreasing weights to past observations, giving more importance to recent data.
+  - **Prophet**: A forecasting tool developed by Facebook that handles seasonality, holidays, and missing data.
```
Comment (sladenheim) [inline ds701_book/26-TimeSeries.qmd:1314]: I think we should drop this. It seems too random to throw it in there. We could mention it as a comment that there are other forecasting tools that are outside the scope of this lecture.

File: ds701_book/26-TimeSeries.qmd (hunk 13 above)
Comment (trgardos) [reply ds701_book/26-TimeSeries.qmd:714]: It is a relative link which should work when the website is published, but I can't test it till then.

File: ds701_book/26-TimeSeries.qmd (hunk 16 above)
Comment (trgardos) [reply ds701_book/26-TimeSeries.qmd:1314]: ok. drop which? Prophet?

File: ds701_book/26-TimeSeries.qmd [hunk 17]
```diff
@@ -0,0 +1,1351 @@
...
+- **Confidence Intervals**: A range of values that is likely to contain the true value of the forecast. It provides an estimate of the uncertainty associated with the forecast.
+  - **Calculation**: Confidence intervals are typically calculated using the standard error of the forecast and a critical value from the t-distribution or normal distribution.
+  - **Interpretation**: A 95% confidence interval means that there is a 95% chance that the true value will fall within the interval.
+  
+**Key Points:**
```
[the remark this answers is not in the document]
Comment (trgardos) [reply ds701_book/26-TimeSeries.qmd:1319]: done.

File: ds701_book/26-TimeSeries.qmd (hunk 13 above)
Comment (sladenheim) [reply ds701_book/26-TimeSeries.qmd:714]: Ok. No problem. Just was confused since it didn't work for me just yet. Thanks!

File: ds701_book/26-TimeSeries.qmd (hunk 16 above)
Comment (trgardos) [reply ds701_book/26-TimeSeries.qmd:1315]: dropped

File: ds701_book/26-TimeSeries.qmd (hunk 8 above)
Comment (trgardos) [reply ds701_book/26-TimeSeries.qmd:1169]: good point. i mention that we are not using any. That's the only model in statsmodels but default is `exog=None`.

File: ds701_book/26-TimeSeries.qmd (hunk 4 above)
Comment (trgardos) [reply ds701_book/26-TimeSeries.qmd:1007]: Added a sentence. Feel free to suggest a revision.

File: ds701_book/26-TimeSeries.qmd (hunk 5 above)
Comment (trgardos) [reply ds701_book/26-TimeSeries.qmd:1315]: Exponential smoothing can also be used, but seems the SARIMA family of models does better.
